ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -1031,3 +1056,34 @@ static const char* NotificationTypeToString(PrjFS_NotificationType notificationT
     }
 }
 #endif
+
+static shared_ptr<mutex> CheckoutPathMutex(const string& fullPath)
+{
+    mutex_lock lock(s_inProgressExpansionsMutex);
+    PathToMutexMap::iterator iter = s_inProgressExpansions.find(fullPath);
+    if (iter == s_inProgressExpansions.end())
+    {
+        pair<PathToMutexMap::iterator, bool> newEntry = s_inProgressExpansions.insert(
+            PathToMutexMap::value_type(fullPath, { make_shared<mutex>(), 1 }));
+        assert(newEntry.second);
+        return newEntry.first->second.mutex;
+    }
+    else
+    {
+        iter->second.useCount++;
```

Comment:
In theory, this could happen outside of the global lock (and use interlocked operations instead). How much do you think that would reduce contention during a parallel build that is doing lots of hydrations? Is it worth the trouble?

Reply:
> In theory, this could happen outside of the global lock (and use interlocked operations instead).

I think there is a scenario where that approach would not be thread safe:

Thread 1: Calls `CheckoutPathMutex` and a new mutex is added to the global collection
Thread 2: Calls `CheckoutPathMutex` and release global lock before incrementing `useCount`
Thread 1: Calls `ReturnPathMutex`, useCount drops to 0 and so the mutex is removed from the global collection
Thread 3: Calls `CheckoutPathMutex`, a new mutex is created for the same path

Thread 2 and 3 would now be holding two different mutexes for the same path.

I think we can leave this as-is for now, and if we identify that it's causing perf problems we can look into improvements\alternatives.